host-tools/callisto-lib/src/lib.rs

```diff
@@ -0,0 +1,327 @@
+// SPDX-FileCopyrightText: 2023 Google LLC
+//
+// SPDX-License-Identifier: Apache-2.0
+use bittide_sys::callisto;
+use std::mem::{align_of, size_of};
+
+/// Rust sibling of `Bittide.ClockControl.SpeedChange`.
+#[repr(u32)]
+pub enum SpeedChange {
+    /// Increases clock speed.
+    SpeedUp = 0,
+    /// Decreases clock speed.
+    SlowDown = 1,
+    /// Keeps the clock as it is.
+    NoChange = 2,
+}
+
+// We use a static assertion to verify that the memory layout appears
+// to be as expected for ensuring a consistent encoding at the Haskell
+// side.
+const _: () = assert!(4 == size_of::<SpeedChange>());
+const _: () = assert!(4 == align_of::<SpeedChange>());
+
+/// Rust sibling of
+/// `Bittide.ClockControl.Callisto.Types.ReframingState`.
+#[repr(C)]
+pub enum ReframingState {
+    /// The controller remains in this state until stability has been
+    /// detected.
+    Detect,
+    /// Reframing has taken place. There is nothing more to do.
+    Done,
+    /// The controller remains in this state for the predefined
+    /// number of cycles with the assumption that the elastic buffers
+    /// of all other nodes are sufficiently stable after that time.
+    Wait {
+        /// Stored correction value to be applied at reframing time.
+        target_correction: f32,
+        /// Number of cycles to wait until reframing takes place.
+        cur_wait_time: u32,
+    },
+}
```

Comment:
One thing to be careful with here is that at some point you cast a `*const ()` to a `*const ControlSt`, which contains a `ReframingState`. The issue is that there is no validation that the tag is valid, so you can easily end up with undefined behaviour if the tag is not 0, 1 or 2 and there's no nice way to detect that either. That's why the code in `callisto-rust-split` uses a struct and a manual `u32` tag field to check it's valid.
Since we are the ones that are building the data structure on the Haskell side it's fine, but there should be some comment about this that there is this possibility if it's misused.

Reply:
You don't think that the

```Rust
/// Rust sibling of `Bittide.ClockControl.Callisto.Types.ReframingState`.
```
and the `# Safety` comments for `__c_callisto_rust` are enough?

Reply:
The `# Safety` section only talks about allocations of appriopriate sizes, not about valid values. I guess it's enough, just something that I noticed while reading through the code :woozy_face: